### PR TITLE
[UART] UART implementation

### PR DESCRIPTION
**Pull request by jprestwo**
- Fixed 'then' in callbacks by acquiring the value. Without this
  promises did not work correctly if an object other than
  undefined was used.

Signed-off-by: James Prestwood james.prestwood@intel.com

### CODE REVIEW

File: src/zjs_callbacks.c [hunk 1]
```diff
@@ -214,6 +214,7 @@ int32_t zjs_add_callback_list(jerry_value_t js_func,
         new_cb->js->this = jerry_acquire_value(this);
         new_cb->js->pre = pre;
         new_cb->js->post = post;
+        new_cb->js->this = jerry_acquire_value(this);
```
Comment (grgustaf) [inline src/zjs_callbacks.c]: You reverted this with a later patch so make sure you fix it in this resubmitted verison. :)

**Comment by jprestwo** [comment]
@poussa, could you try this when you have a chance, specifically the script generation. You had said you had a problem with convert.sh after I added the new line / carriage return changes but I am unable to reproduce, and neither is Travis CI. Thanks

**Comment by poussa** [comment]
Will do, on Monday.

**Comment by poussa** [comment]
@jprestwo The convert.sh script is still causing newline handling troubles to me (same as before).

```
  CC      src/zjs_script_gen.o
/home/spoussa/src/iot/zephyr.js/src/zjs_script_gen.c:5:26: error: missing terminating " character [-Werror]
 const char *script_gen = "console.log('Hello, ZJS world!')
                          ^
/home/spoussa/src/iot/zephyr.js/src/zjs_script_gen.c:5:1: error: missing terminating " character
 const char *script_gen = "console.log('Hello, ZJS world!')
 ^
/home/spoussa/src/iot/zephyr.js/src/zjs_script_gen.c:6:1: error: missing terminating " character [-Werror]
 ";
 ^
/home/spoussa/src/iot/zephyr.js/src/zjs_script_gen.c:6:1: error: missing terminating " character
/home/spoussa/src/iot/zephyr.js/src/zjs_script_gen.c:6:1: error: expected expression at end of input
```

The issue is malformed C string to contain the JS app:

```
[spoussa@zephyr.js ((213bc7e...))]$ ./scripts/convert.sh samples/HelloWorld.js foo.c
100% 0x00 ..................................
[spoussa@zephyr.js ((213bc7e...))]$ cat foo.c 
/* This file was auto-generated */

#include "zjs_common.h"

const char *script_gen = "console.log('Hello, ZJS world!')
";
```
There is extra newline & CR at the end of the console.log statement before the closing `";`

```bash
$ bash --version
GNU bash, version 4.3.46(1)-release (x86_64-pc-linux-gnu)
```

**Comment by jprestwo** [comment]
@poussa, I am currently at home and trying to update the Zephyr SDK version in order to build but I did update bash to the exact same version as you and I am not seeing the new line at the end of the generated .c file (when I run the script manually). Once I upgrade the SDK I will actually run a build, but I am thinking this may be some default bash setting that is turned on in your environment. @grgustaf, @jimmy-huang , @brianjjones, could you all please try this PR and see if your seeing the same issue. To upgrade bash I did:
```
sudo apt-get install --only-upgrade bash
```
```
bash --version
GNU bash, version 4.3.46(1)-release (x86_64-pc-linux-gnu)
```

@poussa, maybe you could try running the convert script manually and see if that generates correctly:
```
./scripts/convert.sh samples/HelloWorld.js gen.c
gedit gen.c
```

**Comment by poussa** [comment]
I has to do something with my bash environment. I just can't figure out what it is now. Here are my opts. You can copy&paste the output and see if it breaks for you.

```bash 
[spoussa@~]$ shopt -p
shopt -u autocd
shopt -u cdable_vars
shopt -u cdspell
shopt -u checkhash
shopt -u checkjobs
shopt -s checkwinsize
shopt -s cmdhist
shopt -u compat31
shopt -u compat32
shopt -u compat40
shopt -u compat41
shopt -u compat42
shopt -s complete_fullquote
shopt -u direxpand
shopt -u dirspell
shopt -u dotglob
shopt -u execfail
shopt -s expand_aliases
shopt -u extdebug
shopt -s extglob
shopt -s extquote
shopt -u failglob
shopt -s force_fignore
shopt -u globstar
shopt -u globasciiranges
shopt -u gnu_errfmt
shopt -s histappend
shopt -u histreedit
shopt -u histverify
shopt -u hostcomplete
shopt -u huponexit
shopt -s interactive_comments
shopt -u lastpipe
shopt -u lithist
shopt -u login_shell
shopt -u mailwarn
shopt -u no_empty_cmd_completion
shopt -u nocaseglob
shopt -u nocasematch
shopt -u nullglob
shopt -s progcomp
shopt -s promptvars
shopt -u restricted_shell
shopt -u shift_verbose
shopt -s sourcepath
shopt -u xpg_echo
```

**Comment by grgustaf** [comment]
Please rebase to remove the three old patches (original patch, "bad merge" fixup, and revert)

**Comment by grgustaf** [comment]
+1, works in qemu and A101.

**Pull request merged**